### ISSUE TITLE
Rule proposal: Warn on functions with consecutive parameters of the same type

### ISSUE THREAD

**Issue by danvk** [typescript-eslint/typescript-eslint#8608]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

When a function takes multiple consecutive parameters of the same type, it's easy to get the order wrong.

### Fail Cases

```TypeScript
function saveUser(name: string, email: string) {
  // ...
}

saveUser('user@example.com', 'Example User'); // backwards, no error!
```

### Pass Cases

```TypeScript
function saveUser(user: User) {
  // ...
}

// or introduce distinct types:

function saveUser(name: string, email: Email) {
  // ...
}
```

### Additional Info

This example comes from https://twitter.com/housecor/status/1764309244554219825

Similar advice appears in _Effective TypeScript_ second edition, Item 38: Avoid Repeated Parameters of the Same Type

There are exceptions to this rule that would be difficult to detect automatically:

- If the function is commutative then order of the arguments is irrelevant and there's no problem. Example: `max(a, b)`
- If there's a "natural" order to the parameters then there may not be ambiguity. Example: `array.slice(start, stop)`

Warning on 3+ consecutive parameters of the same type seems more likely to produce only true positives, but maybe this could be an option.

**Comment by bitjson** [comment]
💯 This would be very useful.

It seems like this could reasonably be made an option of the `eslint/max-params` rule, maybe a `maximumParametersPerType` with a default of `2`? That default value would avoid the mentioned exceptions. 

To validate, the rule would increment a counter for each fully resolved type it encounters in the function definition's parameters, failing if any counter exceeds the maximum.

I'll add that I'd also like to prevent non-consecutive, like-typed parameters in my codebases, e.g.:

With a function signature like `renderTemplate(type: string, version: number, content: string): string`, it often seems just as plausible in code review that the locations of the `type` and `content` should be reversed. 

I'd like eslint to fail `renderTemplate` and recommend that the developer instead use an object parameter, so I'd always set `maximumParametersPerType` to `1`.

**Comment by Josh-Cena** [comment]
@bitjson Your suggestion about non-consecutive but semantically homologous arguments is good. However max-params is a core rule, not our rule, and this proposal is far from a natural extension of that rule, so it will surely be its own thing. I do think it's quite hard to tell what a sensible heuristic for "confusing parameters" would be.

**Comment by bradzacher** [comment]
Personally I'm all for this rule existing, but I don't think this is something that belongs in our general-purpose plugin.

It is ultimately a stylistic rule where most cases can likely be covered with pure AST checks. eg the main cases being solved for are going to be the base primitives (string, number, boolean) used in a direct, non-union style.
You would need type-info for all cases but my gut says that you're talking the last 0.1% of cases need type info.

This is also something that likely is covered mostly by [editor inlays](https://code.visualstudio.com/docs/typescript/typescript-editing#_inlay-hints). Parameter name inlays go a long way to helping reveal such problems when authoring usages of functions. Eg typescript will show an inlay if you use a raw value or if your variable name does not match the parameter name.
Whilst it isn't lint-time enforcement that declarations are correct - it does go a long way to help with the authoring experience.

For now let's see how much community engagement this garners.

**Comment by bradzacher** [comment]
5 months later and this request has received just 5 community reactions -- even with a tweet linking to it.
As per our [workflow guidelines](https://typescript-eslint.io/maintenance/issues#community-engagement-evaluation) - we are taking this as signal that this check is not important to the broader community and are thus rejecting the proposal. Thanks for filing!
